### PR TITLE
Meter static types string conversion

### PR DESCRIPTION
**Pull request by SupunS**
Work towards https://github.com/dapperlabs/cadence-private-issues/issues/47

## Description

Depends on https://github.com/onflow/cadence/pull/1598

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Review by robert-e-davidson3 — APPROVED** [review]

File: runtime/common/metering.go [hunk 1]
```diff
@@ -147,6 +147,17 @@ var (
...
+	LinkValueStringMemoryUsage             = NewRawStringMemoryUsage(len("Link<>()"))
+
+	// Static types string representations
+
+	VariableSizedStaticTypeMemoryUsage = NewRawStringMemoryUsage(2)  // []
+	DictionaryStaticTypeMemoryUsage    = NewRawStringMemoryUsage(4)  // {: }
+	OptionalStaticTypeMemoryUsage      = NewRawStringMemoryUsage(1)  // ?
+	AuthReferenceStaticTypeMemoryUsage = NewRawStringMemoryUsage(5)  // auth&
+	ReferenceStaticTypeMemoryUsage     = NewRawStringMemoryUsage(1)  // &
+	CapabilityStaticTypeMemoryUsage    = NewRawStringMemoryUsage(12) // Capability<>
```
Comment (dsainati1) [inline runtime/common/metering.go:160]: These names are kind of misleading; can these be changed to reflect the fact that they are strings and not types themselves? 
```suggestion
	VariableSizedStaticTypeStringMemoryUsage = NewRawStringMemoryUsage(2)  // []
	DictionaryStaticTypeStringMemoryUsage    = NewRawStringMemoryUsage(4)  // {: }
	OptionalStaticTypeStringMemoryUsage      = NewRawStringMemoryUsage(1)  // ?
	AuthReferenceStaticTypeStringMemoryUsage = NewRawStringMemoryUsage(5)  // auth&
	ReferenceStaticTypeStringMemoryUsage     = NewRawStringMemoryUsage(1)  // &
	CapabilityStaticTypeStringMemoryUsage    = NewRawStringMemoryUsage(12) // Capability<>
```

File: runtime/interpreter/statictype.go [hunk 2]
```diff
@@ -196,6 +221,13 @@ func (t VariableSizedStaticType) String() string {
 	return fmt.Sprintf("[%s]", t.Type)
 }
 
+func (t VariableSizedStaticType) MeteredString(memoryGauge common.MemoryGauge) string {
+	common.UseMemory(memoryGauge, common.VariableSizedStaticTypeMemoryUsage)
```
Comment (dsainati1) [inline runtime/interpreter/statictype.go:225]:
```suggestion
	common.UseMemory(memoryGauge, common.VariableSizedStaticTypeStringMemoryUsage)
```

File: runtime/interpreter/statictype.go [hunk 3]
```diff
@@ -286,6 +332,15 @@ func (t DictionaryStaticType) String() string {
 	return fmt.Sprintf("{%s: %s}", t.KeyType, t.ValueType)
 }
 
+func (t DictionaryStaticType) MeteredString(memoryGauge common.MemoryGauge) string {
+	common.UseMemory(memoryGauge, common.DictionaryStaticTypeMemoryUsage)
```
Comment (dsainati1) [inline runtime/interpreter/statictype.go:336]:
```suggestion
	common.UseMemory(memoryGauge, common.DictionaryStaticTypeStringMemoryUsage)
```

File: runtime/interpreter/statictype.go [hunk 4]
```diff
@@ -323,6 +378,13 @@ func (t OptionalStaticType) String() string {
 	return fmt.Sprintf("%s?", t.Type)
 }
 
+func (t OptionalStaticType) MeteredString(memoryGauge common.MemoryGauge) string {
+	common.UseMemory(memoryGauge, common.OptionalStaticTypeMemoryUsage)
```
Comment (dsainati1) [inline runtime/interpreter/statictype.go:382]:
```suggestion
	common.UseMemory(memoryGauge, common.OptionalStaticTypeStringMemoryUsage)
```

File: runtime/interpreter/statictype.go [hunk 5]
```diff
@@ -435,6 +516,23 @@ func (t ReferenceStaticType) String() string {
 	return fmt.Sprintf("%s&%s", auth, t.BorrowedType)
 }
 
+func (t ReferenceStaticType) MeteredString(memoryGauge common.MemoryGauge) string {
+	if t.Authorized {
+		common.UseMemory(memoryGauge, common.AuthReferenceStaticTypeMemoryUsage)
```
Comment (dsainati1) [inline runtime/interpreter/statictype.go:521]:
```suggestion
		common.UseMemory(memoryGauge, common.AuthReferenceStaticTypeStringMemoryUsage)
```

File: runtime/interpreter/statictype.go [hunk 6]
```diff
@@ -435,6 +516,23 @@ func (t ReferenceStaticType) String() string {
...
+	} else {
+		common.UseMemory(memoryGauge, common.ReferenceStaticTypeMemoryUsage)
```
Comment (dsainati1) [inline runtime/interpreter/statictype.go:523]:
```suggestion
		common.UseMemory(memoryGauge, common.ReferenceStaticTypeStringMemoryUsage)
```

File: runtime/interpreter/statictype.go [hunk 7]
```diff
@@ -477,6 +575,17 @@ func (t CapabilityStaticType) String() string {
 	return "Capability"
 }
 
+func (t CapabilityStaticType) MeteredString(memoryGauge common.MemoryGauge) string {
+	common.UseMemory(memoryGauge, common.CapabilityStaticTypeMemoryUsage)
```
Comment (dsainati1) [inline runtime/interpreter/statictype.go:579]:
```suggestion
	common.UseMemory(memoryGauge, common.CapabilityStaticTypeStringMemoryUsage)
```

File: runtime/common/metering.go (hunk 1 above)
Comment (SupunS) [reply runtime/common/metering.go]: That makes sense! Added with f444896a1f5b9f0aaba562287d4c021c4e63df6f

**Review by dsainati1 — APPROVED** [review]

File: runtime/interpreter/primitivestatictype.go [hunk 8]
```diff
@@ -39,6 +43,18 @@ func (t PrimitiveStaticType) Equal(other StaticType) bool {
...
+func (t PrimitiveStaticType) MeteredString(memoryGauge common.MemoryGauge) string {
+	if str, ok := PrimitiveStaticTypes[t]; ok {
+		common.UseMemory(memoryGauge, common.NewRawStringMemoryUsage(len(str)))
+		return str
+	}
+
+	l := 21 + OverEstimateIntStringLength(int(t))
+	common.UseMemory(memoryGauge, common.NewRawStringMemoryUsage(l))
+
+	return "PrimitiveStaticType(" + strconv.FormatInt(int64(t), 10) + ")"
```
Comment (turbolent) [inline runtime/interpreter/primitivestatictype.go:55]: Maybe calculate the magic number from the constant part of the formatted string, e.g. 
```go
const primitiveStaticTypePrefix = "PrimitiveStaticType"
// + 2 for parentheses
var primitiveStaticTypeConstantLength = len(primitiveStaticTypePrefix) + 2
```
and then `return fmt.Sprintf("%s(%d)", primitiveStaticTypePrefix, t)`

File: runtime/interpreter/statictype.go [hunk 9]
```diff
@@ -98,6 +99,16 @@ func (t CompositeStaticType) String() string {
 	return string(t.TypeID)
 }
 
+func (t CompositeStaticType) MeteredString(memoryGauge common.MemoryGauge) string {
+	if t.Location == nil {
+		common.UseMemory(memoryGauge, common.NewRawStringMemoryUsage(len(t.QualifiedIdentifier)))
+	} else {
+		common.UseMemory(memoryGauge, common.NewRawStringMemoryUsage(len(t.TypeID)))
+	}
```
Comment (turbolent) [inline runtime/interpreter/statictype.go:107]: Maybe factor out the common parts, the calls to construct the usage and the use of the usage:

```suggestion
	var amount int
	if t.Location == nil {
		amount = len(t.QualifiedIdentifier)
	} else {
		amount = len(t.TypeID)
	}
	common.UseMemory(memoryGauge, common.NewRawStringMemoryUsage(amount))
```

File: runtime/interpreter/statictype.go [hunk 10]
```diff
@@ -244,6 +269,20 @@ func (t ConstantSizedStaticType) String() string {
 	return fmt.Sprintf("[%s; %d]", t.Type, t.Size)
 }
 
+func (t ConstantSizedStaticType) MeteredString(memoryGauge common.MemoryGauge) string {
+	// n - for size
+	// 2 - for open and close bracket.
+	// 1 - for space
+	// 1 - for semicolon
```
Comment (turbolent) [inline runtime/interpreter/statictype.go:276]: 👍

**Review by turbolent — APPROVED** [review]
Nice!

File: runtime/interpreter/primitivestatictype.go (hunk 8 above)
Comment (SupunS) [reply runtime/interpreter/primitivestatictype.go]: Added in 49b9ebf

File: runtime/interpreter/statictype.go (hunk 9 above)
Comment (SupunS) [reply runtime/interpreter/statictype.go]: Added in 49b9ebf

**Review by turbolent — APPROVED** [review]

**Pull request merged**